### PR TITLE
Keywords and Predicates in Traversals and Protocol / ResultSet Error Updates

### PR DESCRIPTION
**Pull request by lyndonbauto**
*Issue #, if available:*
AN-1034

*Description of changes:*
Added keyword and predicates to traversals. Testing here is limited but will be covered by incoming Cucumber test suite.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gremlin-go/driver/traversal.go [hunk 1]
```diff
@@ -96,3 +96,250 @@ func (t *Traversal) Iterate() (*Traversal, <-chan bool, error) {
...
+	Id_   T = "id_"
+	Key   T = "key"
+	Value T = "value"
+)
+
+type Operator string
+
+const (
+	Sum     Operator = "sum_"
+	Sum_    Operator = "sum"
```
Comment (lyndonbauto) [inline gremlin-go/driver/traversal.go]:
```suggestion
	Sum     Operator = "sum"
	Sum_    Operator = "sum_"
```

File: gremlin-go/driver/graphBinary.go [hunk 2]
```diff
@@ -726,6 +785,30 @@ func (serializer *graphBinaryTypeSerializer) getSerializerToWrite(val interface{
...
+	case T:
+		return &graphBinaryTypeSerializer{dataType: TType, writer: enumWriter, logHandler: serializer.logHandler}, nil
+	case Barrier:
+		return &graphBinaryTypeSerializer{dataType: BarrierType, writer: enumWriter, logHandler: serializer.logHandler}, nil
+	case Scope:
+		return &graphBinaryTypeSerializer{dataType: ScopeType, writer: enumWriter, logHandler: serializer.logHandler}, nil
+	case Predicate:
+		return &graphBinaryTypeSerializer{dataType: PType, writer: pWriter, logHandler: serializer.logHandler}, nil
+	case TextPredicate:
+		return &graphBinaryTypeSerializer{dataType: TextPType, writer: textPWriter, logHandler: serializer.logHandler}, nil
```
Comment (L0Lmaker) [inline gremlin-go/driver/graphBinary.go:811]: Same as previous comment.

File: gremlin-go/driver/graphBinary.go [hunk 3]
```diff
@@ -646,6 +658,53 @@ const (
...
+	}
+
+	for pValue := range v.values {
+		_, err := typeSerializer.write(pValue, buffer)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buffer.Bytes(), err
+}
```
Comment (L0Lmaker) [inline gremlin-go/driver/graphBinary.go:706]: Do these need readers as well?

File: gremlin-go/driver/gorillaTransporter.go [hunk 4]
```diff
@@ -68,16 +73,47 @@ func (transporter *gorillaTransporter) Write(data []byte) (err error) {
...
 	if transporter.connection == nil {
-		err = transporter.Connect()
+		err := transporter.Connect()
 		if err != nil {
-			return
+			return nil, err
+		}
+	}
+	transporter.connection.SetPongHandler(func(string) error {
+		err := transporter.connection.SetReadDeadline(time.Now().Add(time.Second))
```
Comment (L0Lmaker) [inline gremlin-go/driver/gorillaTransporter.go:84]: `time.Now().Add(time.Second)`
Is this a from a convention or rule?

File: gremlin-go/driver/graphBinary.go (hunk 3 above)
Comment (lyndonbauto) [reply gremlin-go/driver/graphBinary.go:706]: No we do not accept predicates, we only write them.

File: gremlin-go/driver/graphBinary.go (hunk 2 above)
Comment (lyndonbauto) [reply gremlin-go/driver/graphBinary.go:811]: No need, as a client we only send these we do not accept them.

File: gremlin-go/driver/traversal.go [hunk 5]
```diff
@@ -86,13 +86,12 @@ func (t *Traversal) Iterate() (*Traversal, <-chan bool, error) {
 		defer close(r)
 
 		// Force waiting until complete.
-		_ = res.All()
+		_, _ = res.All()
 		r <- true
 	}()
 
 	return t, r, nil
 }
```
Comment (lyndonbauto) [inline gremlin-go/driver/traversal.go]:
```suggestion
}

```

File: gremlin-go/driver/gorillaTransporter.go [hunk 6]
```diff
@@ -68,16 +73,47 @@ func (transporter *gorillaTransporter) Write(data []byte) (err error) {
...
+		if err == nil {
+			return bytes, nil
+		}
+		failureCount += 1
+		if failureCount > maxFailCount {
+			return nil, errors.New(fmt.Sprintf("failed to read from socket more than %d times", maxFailCount))
+		}
+
+		// Try pinging server.
+		err = transporter.connection.SetWriteDeadline(time.Now().Add(500 * time.Millisecond))
```
Comment (simonz-bq) [inline gremlin-go/driver/gorillaTransporter.go:108]: In general I'm not sure if a static timeout for something that can be out of a user's control to be a good idea, but this is just to bring it to attention.

Exponential backoff is probably a good idea

File: gremlin-go/driver/traversal.go [hunk 7]
```diff
@@ -120,3 +119,248 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+	Without(args ...interface{}) Predicate
+	// And Predicate returns a Predicate composed of two predicates (logical AND of them).
+	And(args ...interface{}) Predicate
+	// Or Predicate returns a Predicate composed of two predicates (logical OR of them).
+	Or(args ...interface{}) Predicate
+}
+
+var P *p = &p{}
+
+func newP(operator string, values []interface{}) Predicate {
```
Comment (simonz-bq) [inline gremlin-go/driver/traversal.go]: Why do we need this?

File: gremlin-go/driver/gorillaTransporter.go (hunk 4 above)
Comment (lyndonbauto) [reply gremlin-go/driver/gorillaTransporter.go:84]: It just seemed like a good amount of time

File: gremlin-go/driver/traversal.go [hunk 8]
```diff
@@ -120,3 +119,248 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+
+func newPWithP(operator string, pp p, values []interface{}) Predicate {
+	pSlice := make([]interface{}, len(values)+1)
+	copy(pSlice, values)
+	pSlice[len(pSlice)-1] = pp
+	return &p{operator: operator, values: pSlice}
+}
+
+func (_ *p) Between(args ...interface{}) Predicate {
+	return newP("between", args)
```
Comment (simonz-bq) [inline gremlin-go/driver/traversal.go]: Note for all of these args they aren't expanded. I'm not sure if this is intentional? ```args...```

File: gremlin-go/driver/traversal.go [hunk 9]
```diff
@@ -120,3 +119,248 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+	// StartingWith TextPredicate determines if a string starts with a given value.
+	StartingWith(args ...interface{}) TextPredicate
+}
+
+type textP p
+
+var TextP *textP = &textP{}
+
+func newTextP(operator string, values []interface{}) TextPredicate {
+	return &textP{operator: operator, values: values}
```
Comment (simonz-bq) [inline gremlin-go/driver/traversal.go:352]: Same wiht this constructor?

File: gremlin-go/driver/traversal.go [hunk 10]
```diff
@@ -120,3 +119,248 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+}
+
+func (_ *textP) Containing(args ...interface{}) TextPredicate {
+	return newTextP("containing", args)
```
Comment (simonz-bq) [inline gremlin-go/driver/traversal.go]: Ditto here with the args.

File: gremlin-go/driver/gorillaTransporter.go (hunk 6 above)
Comment (lyndonbauto) [reply gremlin-go/driver/gorillaTransporter.go:108]: I will make a ticket to make it configurable.

File: gremlin-go/driver/traversal.go (hunk 7 above)
Comment (lyndonbauto) [reply gremlin-go/driver/traversal.go]: Updated, it is required.

**Review by simonz-bq — APPROVED** [review]

File: gremlin-go/driver/traversal.go [hunk 11]
```diff
@@ -120,3 +120,258 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+	NotEndingWith(args ...interface{}) TextPredicate
+	// NotStartingWith TextPredicate determines if a string does not start with a given value.
+	NotStartingWith(args ...interface{}) TextPredicate
+	// StartingWith TextPredicate determines if a string starts with a given value.
+	StartingWith(args ...interface{}) TextPredicate
+}
+
+type textP p
+
+var TextP *textP = &textP{}
```
Comment (simonz-bq) [inline gremlin-go/driver/traversal.go]: This should be var TextP Textpredicate = &textP{}

File: gremlin-go/driver/traversal.go [hunk 12]
```diff
@@ -120,3 +120,258 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+	Within(args ...interface{}) Predicate
+	// Without Predicate determines if value is not within the specified
+	Without(args ...interface{}) Predicate
+	// And Predicate returns a Predicate composed of two predicates (logical AND of them).
+	And(args ...interface{}) Predicate
+	// Or Predicate returns a Predicate composed of two predicates (logical OR of them).
+	Or(args ...interface{}) Predicate
+}
+
+var P *p = &p{}
```
Comment (simonz-bq) [inline gremlin-go/driver/traversal.go]: This should be `var P Predicate = &p{}`

File: gremlin-go/driver/traversal.go (hunk 11 above)
Comment (lyndonbauto) [reply gremlin-go/driver/traversal.go]:
```suggestion
var TextP TextPredicate = &textP{}
```

File: gremlin-go/driver/traversal.go (hunk 12 above)
Comment (lyndonbauto) [reply gremlin-go/driver/traversal.go]:
```suggestion
var P Predicate = &p{}
```

**Review by simonz-bq — APPROVED** [review]

File: gremlin-go/driver/traversal.go [hunk 13]
```diff
@@ -120,3 +120,258 @@ func (t *Traversal) getResults() (ResultSet, error) {
...
+const (
+	NormSack Barrier = "normSack"
+)
+
+type Cardinality string
+
+const (
+	Single Cardinality = "single"
+	List   Cardinality = "list_"
+	Set    Cardinality = "set_"
```
Comment (lyndonbauto) [inline gremlin-go/driver/traversal.go]: S
```suggestion
	Set_    Cardinality = "set_"
```

Comment (lyndonbauto) [inline gremlin-go/driver/traversal.go]:
```suggestion
	List   Cardinality = "list"
	Set_    Cardinality = "set"
```

**Pull request merged**